### 7337 Natural Language Processing Home Work 2 Paritosh Rai

#### 1.	In Python, create a method for scoring the vocabulary size of a text, and normalize the score from 0 to 1. It does not matter what method you use for normalization as long as you explain it in a short paragraph. (Various methods will be discussed in the live session.)

In [45]:
# import NLTX and othee libraries
import nltk
# urllib to get info from url. 
import urllib.request
from urllib import request
#from __future__ import division
import numpy as np
from sklearn import preprocessing
from sklearn.preprocessing import minmax_scale
import time
import re
import os
from bs4 import BeautifulSoup
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from string import punctuation
import matplotlib as plt
from nltk.tokenize import RegexpTokenizer
from IPython.display import Image
from IPython.core.display import HTML 
import requests
import urllib.request
#from nltk import word_tokenize
# Ref: https://www.guru99.com/accessing-internet-data-with-python.html#:~:text=1%20Call%20the%20read%20function%20on%20the%20webURL,It%20will%20print%20the%20data%20into%20HTML%20format


## Creat a Function to Read url

In [46]:
#Function to read data from url
def read_url(url):
    text = urllib.request.urlopen(url)
    data = text.read().decode('utf8')
    #print(data)
    return data

### Create a function to Calculate Word Count

In [47]:
#define a function to count number of words in the document 
def word_count(data):
    #we will first strip out all non-alpha characters;
    #then we will convert the text to lower case;
    #then we will count the distinct words
    words = nltk.word_tokenize(data)
    words=[word.casefold() for word in words if word.isalpha()] # ignore cases when comparing
    wc_ln=len(words)
    return wc_ln


### Create a function to Calculate unique Word Count

In [48]:
#define a function to get the vocabulary (unique words) in the document 
def vocabulary_count(data):
    #we will first strip out all non-alpha characters;
    #then we will convert the text to lower case;
    #then we will count the distinct words
    words = nltk.word_tokenize(data)
    words=[word.casefold() for word in words if word.isalpha()] # ignore cases when comparing
    vc_ln=len(set(words)) # count distinct words
    return vc_ln

### Create a Function to Normalize 1

In [49]:
#define a function to normalize the vocabulary (unique words) score in the document
# score is normalized by dividing unique words by total number of words
def normalized_score(data):
    # we will divde unique words by total number of words
    nor_score = vocabulary_count(data)/word_count(data)
    return nor_score


### Create a Function to Normalize 2

In [50]:
### Method to get the Vocab Size and normalize the score

def n_vocab_size(*arg):
    vocab_size = np.array([])
    vocab_size_norm = np.array([])
    
    #### Getting the Vocab Size
    for text in arg:
        vocab_size = np.append(vocab_size,len(set(text)))
    
    #### Normalizing using the formula 
    for vsize in vocab_size:
        vocab_size_norm = np.append(vocab_size_norm,(vsize - vocab_size.min()) /
                                                    (vocab_size.max() - vocab_size.min()))
    
    #### Normalizing using sklearn preprocessing 
    vocab_size_norm_sklearn = minmax_scale(vocab_size, feature_range=(0,1), axis=0)
    
    return(vocab_size,vocab_size_norm,vocab_size_norm_sklearn)


In [51]:
# Pull data for grade 4th, 5th and 6th
#Grade4 book
url='https://www.gutenberg.org/files/49339/49339-0.txt'
data_4=read_url(url)
#Grade5 book
url='https://www.gutenberg.org/files/51000/51000-0.txt'
data_5=read_url(url)
#Grade6 book
url='https://www.gutenberg.org/files/36864/36864-0.txt'
data_6=read_url(url)



In [52]:
# Create Corpus using three grade book
vocab_size = n_vocab_size(data_4, data_5, data_6)
vocab_size

(array([145., 123., 107.]),
 array([1.        , 0.42105263, 0.        ]),
 array([1.        , 0.42105263, 0.        ]))

#### Normalization Method 2

In [53]:
print("Normalized Values by using the formula", *vocab_size[1],sep='\n- ')

Normalized Values by using the formula
- 1.0
- 0.42105263157894735
- 0.0


In [55]:
# define the function to calculate corpus normalization score
def normalized_score_corp(grade):
    # we will divde unique words by total number of words
    nor_score_grade = vocab_size[1][grade-4]
    return nor_score_grade


#### Normilazion Method 3 (using sklearn)

In [56]:
print("Normalized Values by using sklearn", *vocab_size[2],sep='\n- ')

Normalized Values by using sklearn
- 1.0
- 0.42105263157894735
- 0.0


#### School Reading By Grades: Fourth Year

In [57]:
url='https://www.gutenberg.org/files/49339/49339-0.txt'
data_4=read_url(url)
print('Word Count for School Reading By Grades Fourth Year is:', word_count(data_4))
print('') # line space
print ('Vocabulary Count for School Reading By Grades Fourth Year is:',vocabulary_count(data_4))
print('') # line space
print ('Normalized in single doc')
print ('Normalize the score of Vacabulary for School Reading By Grades Fourth Year is: ', normalized_score(data_4))
print('')# line space
print('Normalize using corpus made of the books for Grdae 4th,5th and 6th year')
print ('Normalize 2 the score of Vacabulary for School Reading By Grades Fourth Year is: ', normalized_score_corp(4))

Word Count for School Reading By Grades Fourth Year is: 50603

Vocabulary Count for School Reading By Grades Fourth Year is: 6012

Normalized in single doc
Normalize the score of Vacabulary for School Reading By Grades Fourth Year is:  0.11880718534474241

Normalize using corpus made of the books for Grdae 4th,5th and 6th year
Normalize 2 the score of Vacabulary for School Reading By Grades Fourth Year is:  1.0


Two normalization methods were calculated. 

**First Method:** Normalization was done using the respective grade yearbook.  This method was not very effective when the effort is made to compare documents across the corpus. 

**Second Method:** In the second method, normalization was derived across the corpus of grade years 4,5, and 6. We will compare these documents' difficulty levels with each other, so the normalized value derived from the 2nd method will be used.  

#### School Reading By Grades: Fifth Year

In [58]:

url='https://www.gutenberg.org/files/51000/51000-0.txt'
data_5=read_url(url)
print('Word Count for School Reading By Grades Fifth Year is:', word_count(data_5))
print('') # line space
print ('Vocabulary Count for School Reading By Grades Fifth Year is:',vocabulary_count(data_5))
print('') # line space
print('Normalized in single doc')
print ('Normalize the score of Vacabulary for School Reading By Grades Fifth Year is: ', normalized_score(data_5))
print('') # line space
print('Normalize using corpus made of the books for Grdae 4th,5th and 6th year')
print ('Normalize 2 the score of Vacabulary for School Reading By Grades Fifth Year is: ', normalized_score_corp(5))

Word Count for School Reading By Grades Fifth Year is: 45103

Vocabulary Count for School Reading By Grades Fifth Year is: 6839

Normalized in single doc
Normalize the score of Vacabulary for School Reading By Grades Fifth Year is:  0.15163071192603597

Normalize using corpus made of the books for Grdae 4th,5th and 6th year
Normalize 2 the score of Vacabulary for School Reading By Grades Fifth Year is:  0.42105263157894735


#### School Reading By Grades: Sixth Year¶

In [59]:
url='https://www.gutenberg.org/files/36864/36864-0.txt'
data_6=read_url(url)
print('Word Count for School Reading By Grades Sixth Year is:', word_count(data_6))
print('') # line space
print ('Vocabulary Count for School Reading By Grades Sixth Year is:',vocabulary_count(data_6))
print('') # line space
print('Normalized in single doc')
print ('Normalize the score of Vacabulary for School Reading By Grades Sixth Year is (Normlize 1): ', normalized_score(data_6))
print('') # line space
print('Normalize using corpus mode of the books for Grdae 4th,5th and 6th year')
print ('Normalize 2 the score of Vacabulary for School Reading By Grades Fifth Year is: ', normalized_score_corp(6))

Word Count for School Reading By Grades Sixth Year is: 60017

Vocabulary Count for School Reading By Grades Sixth Year is: 8137

Normalized in single doc
Normalize the score of Vacabulary for School Reading By Grades Sixth Year is (Normlize 1):  0.1355782528283653

Normalize using corpus mode of the books for Grdae 4th,5th and 6th year
Normalize 2 the score of Vacabulary for School Reading By Grades Fifth Year is:  0.0


#### 2.	After consulting section 3.2 in chapter 1 of Bird-Klein, create a method for scoring the long-word vocabulary size of a text, and likewise normalize (and explain) the scoring as in step 1 above.

### Define Function to Count Long Words

In [60]:
#define a function to count long words
# give the name of data and length of long word
def count_long_word(data,n):
    #tokenize the data set
    word = nltk.word_tokenize(data)
    #keep only apphabets
    words=[word.casefold() for word in word if word.isalpha()]
    #remove the duplicate
    words = set(words)
    #find long words
    long_words = [w for w in words if len(w) > n]
    #count number of long words
    lw_len=len(long_words)
    return lw_len
    

### Define Function to Normalize Long Word

In [61]:
#define a function to normalize the long vocabulary (unique words) score in the document
# score is normalized by dividing number of long unique words by total number of unique words
def normalized_longword_score(data,n):
    # we will divde unique words by total number of words
    nor_lw_score = count_long_word(data,n)/vocabulary_count(data)
    return nor_lw_score

#### Long Word Count and Normalize Long Word Grade 6
Word length of 15 was consered as  long word  as Natural Language Processing with Python by Steven Bird, Ewan Klein, Edward Loper in section 1.3

In [62]:
# Long word was assumed to be weith with eight (15) letters
print('Count of Long Vocabulary for the School Reading By Grades Six is:', count_long_word(data_6,15))
print('Normalize Long Vocabulary the of Long vocabulary for the School Reading By Grades Six is:', normalized_longword_score(data_6,15))


Count of Long Vocabulary for the School Reading By Grades Six is: 6
Normalize Long Vocabulary the of Long vocabulary for the School Reading By Grades Six is: 0.000737372496005899


In [63]:
# As longword count was only 6, with word length of 15, after few attempts adjusted the longword length to 12
print('Count of Long Vocabulary for the School Reading By Grades Six is:', count_long_word(data_6,12))
print('Normalize Long Vocabulary the of Long vocabulary for the School Reading By Grades Six is:', normalized_longword_score(data_6,12))


Count of Long Vocabulary for the School Reading By Grades Six is: 89
Normalize Long Vocabulary the of Long vocabulary for the School Reading By Grades Six is: 0.010937692024087502


#### Long Word Count and Normalize Long Word Grade 5

In [64]:
# Long word was assumed to be weith with ten (10) letters
print('Count of Long Vocabulary for the School Reading By Grades five year with word lenght 12:', count_long_word(data_5,12))
print('Count of Long Vocabulary for the School Reading By Grades five uear is:', count_long_word(data_5,11))
print('Normalize Long Vocabulary the of Long vocabulary for the School Reading By Grades five Year is:', normalized_longword_score(data_5,11))


Count of Long Vocabulary for the School Reading By Grades five year with word lenght 12: 56
Count of Long Vocabulary for the School Reading By Grades five uear is: 147
Normalize Long Vocabulary the of Long vocabulary for the School Reading By Grades five Year is: 0.021494370522006142


#### Long Word Count and Normalize Long Word Grade 4

In [65]:
# Long word was assumed to be weith with ten (10) letters
print('Count of Long Vocabulary for the School Reading By Grades four year with word lenght 12:', count_long_word(data_4,12))
print('Count of Long vocabulary for the School Reading By Grades four Year is:', count_long_word(data_4,10))
print('Normalize Long Vocabulary the of Long vocabulary for the School Reading By Grade four Year is:', normalized_longword_score(data_4,10))

Count of Long Vocabulary for the School Reading By Grades four year with word lenght 12: 41
Count of Long vocabulary for the School Reading By Grades four Year is: 246
Normalize Long Vocabulary the of Long vocabulary for the School Reading By Grade four Year is: 0.04091816367265469


The above function was used to find the total number of long words used in grade-level books. The function is created to calculate the Normalized Long Vocabulary. We started with a longword length of 15 as per guidance from  **"Natural Language Processing with Python" by Steven Bird, Ewan Klein, Edward Loper in section 1.3**. However, we could find only six words and normalize value was very low as well. After a few iterations, longword sizes of 12,11 and 10 were chosen for 6th,5th, and a 4th-grade year respectively. 

To normalize the outcome of long words, the count of long words in the respective grade year book was divided by the number of unique words in the respective book. 

It was observed that long word counts were 89, 56, and 41for the 6th,5th, and 4th-grade year, respectively, when word size was 12. The long word count was aligned with expectation. As the grade level goes up, the number of long words is expected to go up. 

Hower to align the difficulty level word count length was reduced by one as grade year decrease. The assumption is that as readers go in higher grade, year-long word length should also increase (how much is debatable, we reduced the count to increase by one in the exercise). However, the function was built with the flexibility to adjust if needed. After a few iterations, longword sizes of 12,11, and 10 were chosen for 6th,5th, and 4th-grade yeara, respectively.

**Conclusion:** Outcome was not very aligned with expectations. We expected longword count and normalized longword to coverage with the adjustment of the length of the longword, but the number of long words and the normalized count went in reverse order.  

#### 3.	Now create a “text difficulty score” by combining the lexical diversity score from homework 1, and your normalized score of vocabulary size and long-word vocabulary size, in equal weighting. Explain what you see when this score is applied to same graded texts you used in homework 1.

In [66]:
# function to calculate Lexical Diversity
def lex_div(data):
    ld = len(data)/len(set(data))
    return ld

In [67]:
# Lexical Diversity for grade 6 years
lex_div6=lex_div(data_6)
lex_div6

3396.2897196261683

In [68]:
# Lexical Diversity for grade 5 years
lex_div5=lex_div(data_5)
lex_div5

2210.878048780488

In [69]:
# Lexical Diversity for grade 4 years
lex_div4=lex_div(data_4)
lex_div4

2168.8137931034485

**Lexical Diversity** is the number of times each vocabulary item appears in the text on average in a given book, text, or doc. Higher the number more the repetition. Based on the Lexical Diversity Score, we can say words are repeated more in the 6th-grade year book than the 5th-grade year book, and the 5th-grade year have more repetition of words than the 4th. However, we expected to see less repetition as the grade goes up as new words will be introduced. 

### Create Function to Calculate Text Difficulty Score
Three factors will be taken into account the difficulty score of the respective Score:

**Lexical Diversity:** Number of times each vocabulary item appears in the text on average

**Normalized Score for vocabulary size:** Long word score count was normalized across the corpus to calculated to bring in uniformity across the three books. 

**Long Word Vaculabury Size:** Long word vocabulary size was adjusted to account for the grade year level. A longword size of 12,11, and 10 was chosen for the 6th,5th, and 4th-grade year, respectively.

**Difficulty Score:** Difficulty Score was calculated taking three factors into account Lexical Diversity, Normalized Score for Vocabulary Size, and Long Word Vaculabury Size. 

Following is the relationship of these measures with the difficulty score:

**Lexical Score:** Higher the Lexical Score, there is more repetition of words, i.e., fewer new words, so the difficulty level is lower, i.e., Lexical score and difficulty level are inversely related. 
 
***Normalized Score for Vocabulary Size:** Vocabe size was normalized across the corpus (i.e., all three grade year books). ThisNormilization was carried out to ensure they are measured on the same scale. Normalized values were higher for grade year 4 (i.e., 1) and lowest for grade year 6 (i.e., 0), so higher value for lower grade. 

**Long Word Vocabulary Size:**  Longer the Vocabulary higher the difficulty score. So Difficulty Score and Vocabulary Size should be directly related. We have normalized the value based on the respective document and adjusted the length based on grade year. Higher the grade year, longer the word. 

**To sumaraize:**

* Difficulty Score => Inversely Propstional( Lexical Diversity)

* Difficulty Score => Directly Propostion (Normalized Score for Vocabulary Size)

* Difficulty Score => Inversely Propstional (Long Word Vocabulary Size)

The difficulty level function was created by subtracting normalized_score_corp(grade) from normalized longword Score. The difference is multiplied by 1,000,000 to deal with decimal values. The product of 1 million and difference is squared, and then square root was applied to eliminate negative values. The numerator was divide by Lexical Diversity. 


In [70]:
# created function to calculate text difficulty score
# function for difficulty score was calculated by multiplying normalized_longeword_score (adjsuted for grader level) with 1,000,000 and devicing them by lexical diversity and 
# long word normilozation were be adjusted by changing the length of the word for respective grades
def text_difficulty_score(data,n,grade):
    text_dif_score = (((normalized_longword_score(data,n) - normalized_score_corp(grade)) *1000000/ lex_div(data))**2)**0.5
    return text_dif_score

In [71]:
# text difficulty level calcualted using same text of eight(8) for all grade levels
# if the length of long words were kept same difficulty level is expected to go up.
# as the books in higher grades will have more diverse words
print('Lexical Diversity of School Reading School Reading By Grades Sixth Year is:', lex_div(data_6))
print('Normalize Long Vocabulary the of Long vocabulary for the School Reading By Grade Sixth Year is:', normalized_longword_score(data_6,8))
print('Text Difficulty Score of text for the School Reading By Grade Sixth Year is:', text_difficulty_score(data_6,12,6))

Lexical Diversity of School Reading School Reading By Grades Sixth Year is: 3396.2897196261683
Normalize Long Vocabulary the of Long vocabulary for the School Reading By Grade Sixth Year is: 0.21334644217770676
Text Difficulty Score of text for the School Reading By Grade Sixth Year is: 3.2204826228109362


In [72]:
# text difficulty level calcualted using same text of eight(8) for all grade levels
# if the length of long words were kept same difficulty level is expected to go up.
# as the books in higher grades will have more diverse words
print('Lexical Diversity of School Reading School Reading By Grades Fiveth Year is:', lex_div(data_5))
print('Normalize Long Vocabulary the of Long vocabulary for the School Reading By Grade Fivth Year is:', normalized_longword_score(data_5,8))
print('Text Difficulty Score of text for the School Reading By Grade Fiveth Year is:', text_difficulty_score(data_5,11,5))

Lexical Diversity of School Reading School Reading By Grades Fiveth Year is: 2210.878048780488
Normalize Long Vocabulary the of Long vocabulary for the School Reading By Grade Fivth Year is: 0.18862406784617633
Text Difficulty Score of text for the School Reading By Grade Fiveth Year is: 180.72379038605771


In [73]:
# text difficulty level calcualted using same text of eight(8) for all grade levels
# if the length of long words were kept same difficulty level is expected to go up.
# as the books in higher grades will have more diverse words
print('Lexical Diversity of School Reading School Reading By Grades Fourth Year is:', lex_div(data_4))
print('Normalize Long Vocabulary the of Long vocabulary for the School Reading By Grade Fourth Year is:', normalized_longword_score(data_4,8))
print('Text Difficulty Score of text for the School Reading By Grade Fourth Year is:', text_difficulty_score(data_4,10,4))

Lexical Diversity of School Reading School Reading By Grades Fourth Year is: 2168.8137931034485
Normalize Long Vocabulary the of Long vocabulary for the School Reading By Grade Fourth Year is: 0.17980705256154358
Text Difficulty Score of text for the School Reading By Grade Fourth Year is: 442.21492844480395


**Conclusion:**
The difficulty score was inconsistent with the expectation. Grader 4 had the highest score of ~442, and Grade 6 was ~3. Whereas Grade 5 was in the middle at 180. 
However, the Lexical Score was also in reverse order, i.e., higher Lexical Score for the higher grade. So created Difficulty Score is incorrect; somehow, it did not align with the given data. 